components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/util/ClientCredentialAuthConfig.java

```diff
@@ -0,0 +1,464 @@
+        }
+    }
+
+    @Override
+    public HttpUriRequest applyAuth(HttpUriRequest request, AuthConfigModel authConfigModel)
+            throws FrameworkException {
+
+        maxRequestAttemptsForAPIEndpointTimeout = ConfigProvider.getInstance().
+                getRequestRetryCount();
+        this.apiAccessTokenCache = APIAccessTokenCache.getInstance();
+        Map<String, Object> properties = authConfigModel.getProperties();
+        validateRequiredProperties(properties);
+
+        setConsumerKey(properties.get(CONSUMER_KEY_VARIABLE_NAME).toString());
+        setConsumerSecret(properties.get(CONSUMER_SECRET_VARIABLE_NAME).toString());
+        setTokenEndpoint(properties.get(TOKEN_ENDPOINT).toString());
+        setScopes(properties.containsKey(SCOPES) ? properties.get(SCOPES).toString() : null);
+
+        String accessToken = getAccessToken();
+        if (accessToken == null) {
+            asyncReturn.accept(authenticationContext, Collections.emptyMap(), OUTCOME_FAIL);
+            if (LoggerUtils.isDiagnosticLogsEnabled()) {
+                DiagnosticLog.DiagnosticLogBuilder diagnosticLogBuilder = new
+                        DiagnosticLog.DiagnosticLogBuilder(Constants.LogConstants.ADAPTIVE_AUTH_SERVICE,
+                        Constants.LogConstants.ActionIDs.RECEIVE_TOKEN);
+                diagnosticLogBuilder.inputParam(Constants.LogConstants.InputKeys.TOKEN_ENDPOINT, getTokenEndpoint())
+                        .configParam(Constants.LogConstants.ConfigKeys.SUPPORTED_GRANT_TYPES,
+                                GRANT_TYPE_CLIENT_CREDENTIALS)
+                        .resultMessage("Failed to retrieve access token for the provided token endpoint.")
+                        .logDetailLevel(DiagnosticLog.LogDetailLevel.APPLICATION)
+                        .resultStatus(DiagnosticLog.ResultStatus.FAILED);
+                LoggerUtils.triggerDiagnosticLogEvent(diagnosticLogBuilder);
+            }
+            LOG.error("Failed to retrieve access token. Aborting request.");
+            throw new FrameworkException("Failed to retrieve access token.");
+        }
+        request.setHeader(AUTHORIZATION, BEARER + accessToken);
+        return request;
+    }
+
+    /**
+     * This method decodes access token and compare its expiry time with the current time to decide whether it's
+     * expired.
+     *
+     * @param accessToken Access token which needs to be evaluated
+     * @return A boolean value indicating whether the token is expired
+     * @throws ParseException {@link ParseException}
+     */
+    private boolean isTokenExpired(String accessToken) throws ParseException {
+
+        SignedJWT decodedToken = SignedJWT.parse(accessToken);
+        Date expiryDate = (Date) decodedToken.getJWTClaimsSet().getClaim(JWT_EXP_CLAIM);
+        LocalDateTime expiryTimestamp = LocalDateTime.ofInstant(expiryDate.toInstant(), ZoneId.systemDefault());
+        return LocalDateTime.now().isAfter(expiryTimestamp);
+    }
```

Comment:
Rather than throwing the exception better handle it here and log it.
That clear outs the functional context separation